File: grind75/Dec30/46.go
func permute(nums []int) [][]int {
    ret := [][]int{}
    n := len(nums)
    visited := make([]int, n)
    var dfs func([]int)
    dfs = func(cur []int) {
        if len(cur) == n {
            ret = append(ret, cur)
            return
        }
        for i := 0; i < n; i++ {
            if visited[i] == 1 {
                continue
            }
            tmp := append([]int{}, cur...)
            tmp = append(tmp, nums[i])
            visited[i] = 1
            dfs(tmp)
            visited[i] = 0
        }
    }
    dfs([]int{})
    return ret
}

File: grind75/Dec30/46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        m = len(nums)
        def dfs(perm, current_i):
            if current_i == m - 1:
                result.append(perm)
                return
            for i in range(current_i, m):
                new_perm = perm.copy()
                new_perm[i], new_perm[current_i] = new_perm[current_i], 
new_perm[i]
                dfs(new_perm, current_i + 1)
        dfs(nums, 0)
        return result
